activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1909,15 +1922,20 @@ def build_with(arel)
         return if with_values.empty?
 
         with_statements = with_values.map do |with_value|
-          build_with_value_from_hash(with_value)
+          case with_value
+          when Arel::Nodes::Cte then with_value
+          when Hash then build_with_value_from_hash(with_value)
+          else
+            raise ArgumentError, "Unsupported argument type: #{with_value} #{with_value.class}"
+          end
```

Comment:
There are three arel node types that have `#to_cte` methods: `As`, `TableAlias`, and `Cte` (which, of course, just returns `self`). I wonder if we should allow any object that supports `#to_cte` here.

Reply:
Though, if we are going to support this, I suppose it would be better to do so in `#process_with_args` instead of here, so that by the time we get to this code, all of those node types would be replaced with `Cte`s.